File: image_classifier.py
#!/usr/bin/env python3
"""
AI Image Classification Service
Classifies images using NVIDIA's Florence-2 model
"""

import os
import sys
import requests
import mimetypes
import zipfile
import json
import tempfile
import argparse
from datetime import datetime
from typing import Optional, Dict, Any

# Global debug flag
DEBUG_MODE = False

def debug_print(message: str):
    """Print debug message only if debug mode is enabled"""
    if DEBUG_MODE:
        print(message)

# Load configuration
try:
    from config import API_KEY, NVAi_URL
except ImportError:
    # Fallback to environment variables
    API_KEY = os.getenv("NVIDIA_API_KEY")
    NVAi_URL = "https://ai.api.nvidia.com/v1/vlm/microsoft/florence-2"

if not API_KEY:
    raise ValueError("API_KEY must be set in config.py or as NVIDIA_API_KEY environment variable")

HEADER_AUTH = f"Bearer {API_KEY}"


class ImageClassifier:
    """Handles image classification using NVIDIA's API"""
    
    def __init__(self, api_key: str = None, api_url: str = None):
        self.api_key = api_key or API_KEY
        self.api_url = api_url or NVAi_URL
        self.header_auth = f"Bearer {self.api_key}"
    
    def upload_asset(self, image_path: str, description: str = "Test Image") -> str:
        """Upload image asset to NVIDIA's service"""
        debug_print(f"DEBUG: Starting asset upload for {image_path}")
        content_type, _ = mimetypes.guess_type(image_path)
        if not content_type or not content_type.startswith('image/'):
            content_type = "image/jpeg"
        debug_print(f"DEBUG: Content type: {content_type}")
        
        try:
            debug_print("DEBUG: Sending authorization request...")
            authorize = requests.post(
                "https://api.nvcf.nvidia.com/v2/nvcf/assets",
                headers={
                    "Authorization": self.header_auth,
                    "Content-Type": "application/json",
                    "accept": "application/json",
                },
                json={"contentType": content_type, "description": description},
                timeout=30,
            )
            
            debug_print(f"DEBUG: Authorization response status: {authorize.status_code}")
            authorize.raise_for_status()
            auth_response = authorize.json()
            debug_print(f"DEBUG: Authorization response keys: {list(auth_response.keys())}")
            upload_url = auth_response["uploadUrl"]
            debug_print(f"DEBUG: Upload URL obtained: {upload_url[:50]}...")

            debug_print("DEBUG: Uploading file...")
            with open(image_path, "rb") as f:
                response = requests.put(
                    upload_url,
                    data=f,
                    headers={
                        "x-amz-meta-nvcf-asset-description": description,
                        "content-type": content_type,
                    },
                    timeout=300,
                )
            debug_print(f"DEBUG: Upload response status: {response.status_code}")
            response.raise_for_status()
            asset_id = str(auth_response["assetId"])
            debug_print(f"DEBUG: Asset uploaded successfully, ID: {asset_id}")
            return asset_id
            
        except requests.RequestException as e:
            debug_print(f"DEBUG: Request exception during upload: {str(e)}")
            if hasattr(e, 'response') and e.response is not None:
                debug_print(f"DEBUG: Response status: {e.response.status_code}")
                debug_print(f"DEBUG: Response text: {e.response.text}")
            raise Exception(f"Failed to upload asset: {str(e)}")
    
    def classify_image(self, image_path: str, task: str = "<CAPTION>") -> str:
        """Classify an image and return the result"""
        try:
            debug_print(f"DEBUG: Starting classification for {image_path} with task {task}")
            asset_id = self.upload_asset(image_path)
            content = f'{task}<img src="data:image/jpeg;asset_id,{asset_id}" />'
            inputs = {"messages": [{"role": "user", "content": content}]}
            debug_print(f"DEBUG: Classification input prepared")

            headers = {
                "Content-Type": "application/json",
                "NVCF-INPUT-ASSET-REFERENCES": asset_id,
                "NVCF-FUNCTION-ASSET-IDS": asset_id,
                "Authorization": self.header_auth,
                "Accept": "application/json",
            }
            debug_print(f"DEBUG: Sending classification request to {self.api_url}")

            response = requests.post(self.api_url, headers=headers, json=inputs)
            debug_print(f"DEBUG: Classification response status: {response.status_code}")
            response.raise_for_status()
            
            content_type = response.headers.get('content-type', '').lower()
            debug_print(f"DEBUG: Response content type: {content_type}")
            
            if 'application/json' in content_type:
                result = response.json()
                debug_print(f"DEBUG: JSON response received")
                return result["choices"][0]["message"]["content"]
            elif 'application/zip' in content_type:
                debug_print(f"DEBUG: ZIP response received, extracting...")
                return self._extract_zip_response(response.content)
            else:
                debug_print(f"DEBUG: Unknown response format: {content_type}")
                return "Classification failed - unknown response format"
                
        except Exception as e:
            debug_print(f"DEBUG: Classification exception: {str(e)}")
            if hasattr(e, 'response') and e.response is not None:
                debug_print(f"DEBUG: Error response status: {e.response.status_code}")
                debug_print(f"DEBUG: Error response text: {e.response.text}")
            raise Exception(f"Classification failed: {str(e)}")
    
    def _extract_zip_response(self, zip_content: bytes) -> str:
        """Extract classification result from ZIP response"""
        with tempfile.TemporaryDirectory() as temp_dir:
            zip_path = os.path.join(temp_dir, "response.zip")
            with open(zip_path, 'wb') as f:
                f.write(zip_content)
            
            with zipfile.ZipFile(zip_path, 'r') as zip_ref:
                zip_ref.extractall(temp_dir)
                files = zip_ref.namelist()
                
                response_file = None
                for file in files:
                    if file.endswith('.response'):
                        response_file = file
                        break
                
                if response_file:
                    response_path = os.path.join(temp_dir, response_file)
                    with open(response_path, 'r', encoding='utf-8') as f:
                        json_content = json.load(f)
                        content = json_content["choices"][0]["message"]["content"]
                        if content.startswith('<CAPTION>'):
                            content = content[9:]
                        return content
        
        return "Classification failed - could not extract result from ZIP"


def save_classification_result(image_path: str, classification_result: str, output_dir: str) -> str:
    """Save classification result to JSON file"""
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    image_filename = os.path.basename(image_path)
    timestamp = datetime.now().isoformat()
    
    result_data = {
        "timestamp": timestamp,
        "image_file": image_filename,
        "image_path": image_path,
        "classification": classification_result,
        "processed_at": timestamp
    }
    
    base_name = os.path.splitext(image_filename)[0]
    json_filename = f"{base_name}_classification.json"
    json_path = os.path.join(output_dir, json_filename)
    
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(result_data, f, indent=2, ensure_ascii=False)
    
    return json_path


def main():
    """Main function for command-line usage"""
    parser = argparse.ArgumentParser(description="Classify images using NVIDIA AI")
    parser.add_argument("image_path", help="Path to the image file to classify")
    parser.add_argument("output_dir", help="Directory to save classification results")
    parser.add_argument("--task", default="<CAPTION>", help="Classification task (default: <CAPTION>)")
    parser.add_argument("--debug", action="store_true", help="Enable debug output")
    
    args = parser.parse_args()
    
    # Global debug flag
    global DEBUG_MODE
    DEBUG_MODE = args.debug
    
    try:
        # Debug logging
        debug_print(f"DEBUG: Python script started with args: {args}")
        debug_print(f"DEBUG: Current working directory: {os.getcwd()}")
        debug_print(f"DEBUG: Python version: {sys.version}")
        debug_print(f"DEBUG: Image path: {args.image_path}")
        debug_print(f"DEBUG: Output directory: {args.output_dir}")
        
        # Check API key
        if not API_KEY:
            raise ValueError("API_KEY is not set. Check config.py or environment variables.")
        debug_print(f"DEBUG: API key is present (length: {len(API_KEY)})")
        
        # Validate inputs
        if not os.path.exists(args.image_path):
            raise FileNotFoundError(f"Image file not found: {args.image_path}")
        
        # Check file size
        file_size = os.path.getsize(args.image_path)
        debug_print(f"DEBUG: Image file size: {file_size} bytes")
        if file_size == 0:
            raise ValueError("Image file is empty")
        
        # Test requests module
        debug_print("DEBUG: Testing requests module...")
        import requests
        debug_print(f"DEBUG: Requests module loaded successfully, version: {requests.__version__}")
        
        # Initialize classifier and process image
        debug_print("DEBUG: Initializing classifier...")
        classifier = ImageClassifier()
        debug_print("DEBUG: Starting classification...")
        classification = classifier.classify_image(args.image_path, args.task)
        debug_print(f"DEBUG: Classification completed: {classification[:100]}...")
        
        result_path = save_classification_result(args.image_path, classification, args.output_dir)
        debug_print(f"DEBUG: Result saved to: {result_path}")
        
        # Output result as JSON for Node.js to parse
        result = {
            "success": True,
            "classification": classification,
            "result_file": result_path,
            "image_file": os.path.basename(args.image_path)
        }
        print("CLASSIFICATION_RESULT:" + json.dumps(result))
        
    except Exception as e:
        print(f"ERROR: {str(e)}")
        print(f"ERROR: Exception type: {type(e).__name__}")
        import traceback
        print(f"ERROR: Traceback:\n{traceback.format_exc()}")
        
        result = {
            "success": False,
            "error": str(e),
            "error_type": type(e).__name__,
            "image_file": os.path.basename(args.image_path) if args.image_path and os.path.exists(args.image_path) else "unknown"
        }
        print("CLASSIFICATION_RESULT:" + json.dumps(result))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: imageClassifier.ts
#!/usr/bin/env node

/**
 * AI Image Classification Service
 * Classifies images using NVIDIA's Florence-2 model
 */

import fs from "fs/promises";
import path from "path";
import { createReadStream, existsSync, statSync } from "fs";
import { lookup } from "mime-types";
import fetch from "node-fetch";
import { createWriteStream } from "fs";
import { pipeline } from "stream/promises";
import { createGunzip } from "zlib";
import AdmZip from "adm-zip";

// Types
interface ClassificationResult {
  success: boolean;
  classification?: string;
  result_file?: string;
  image_file?: string;
  error?: string;
  error_type?: string;
}

interface AssetUploadResponse {
  assetId: string;
  uploadUrl: string;
}

interface ClassificationResponse {
  choices: Array<{
    message: {
      content: string;
    };
  }>;
}

interface ResultData {
  timestamp: string;
  image_file: string;
  image_path: string;
  classification: string;
  processed_at: string;
}

// Configuration
let API_KEY: string;
let NVAi_URL: string;
let ASSET_UPLOAD_URL: string;

try {
  // Try to load from config file
  const config = require("./config.js");
  API_KEY = config.nvidia.apiKey;
  NVAi_URL = config.nvidia.apiUrl;
  ASSET_UPLOAD_URL = config.nvidia.assetUploadUrl;
} catch (error) {
  console.error(
    "Failed to load config.js:",
    error instanceof Error ? error.message : String(error)
  );
  // Fallback to environment variables
  API_KEY = process.env.NVIDIA_API_KEY || "";
  NVAi_URL = "https://ai.api.nvidia.com/v1/vlm/microsoft/florence-2";
  ASSET_UPLOAD_URL = "https://api.nvcf.nvidia.com/v2/nvcf/assets";
}

if (!API_KEY) {
  throw new Error(
    "API_KEY must be set in config.js or as NVIDIA_API_KEY environment variable"
  );
}

// Global debug flag
let DEBUG_MODE = true; // Enable debug for troubleshooting

function debugPrint(message: string): void {
  if (DEBUG_MODE) {
    console.log(message);
  }
}

export class ImageClassifier {
  private apiKey: string;
  private apiUrl: string;
  private assetUploadUrl: string;
  private headerAuth: string;

  constructor(apiKey?: string, apiUrl?: string, assetUploadUrl?: string) {
    this.apiKey = apiKey || API_KEY;
    this.apiUrl = apiUrl || NVAi_URL;
    this.assetUploadUrl = assetUploadUrl || ASSET_UPLOAD_URL;
    this.headerAuth = `Bearer ${this.apiKey}`;
  }

  /**
   * Upload image asset to NVIDIA's service
   */
  async uploadAsset(
    imagePath: string,
    description = "Test Image"
  ): Promise<string> {
    debugPrint(`DEBUG: Starting asset upload for ${imagePath}`);

    let contentType = lookup(imagePath) || "image/jpeg";
    if (!contentType.startsWith("image/")) {
      contentType = "image/jpeg";
    }
    debugPrint(`DEBUG: Content type: ${contentType}`);

    try {
      debugPrint("DEBUG: Sending authorization request...");
      debugPrint(`DEBUG: Using asset upload URL: ${this.assetUploadUrl}`);
      debugPrint(
        `DEBUG: Request headers: ${JSON.stringify({
          Authorization: "Bearer [REDACTED]",
          "Content-Type": "application/json",
          accept: "application/json",
        })}`
      );
      debugPrint(
        `DEBUG: Request body: ${JSON.stringify({
          contentType,
          description,
        })}`
      );

      const authorizeResponse = await fetch(this.assetUploadUrl, {
        method: "POST",
        headers: {
          Authorization: this.headerAuth,
          "Content-Type": "application/json",
          accept: "application/json",
        },
        body: JSON.stringify({
          contentType,
          description,
        }),
      });

      debugPrint(
        `DEBUG: Authorization response status: ${authorizeResponse.status}`
      );
      debugPrint(
        `DEBUG: Authorization response headers: ${JSON.stringify(
          Object.fromEntries(authorizeResponse.headers.entries())
        )}`
      );

      if (!authorizeResponse.ok) {
        const responseText = await authorizeResponse.text();
        debugPrint(`DEBUG: Authorization error response body: ${responseText}`);
        throw new Error(
          `Authorization failed: ${authorizeResponse.status} ${authorizeResponse.statusText}. Response: ${responseText}`
        );
      }

      const authResponse =
        (await authorizeResponse.json()) as AssetUploadResponse;
      debugPrint(
        `DEBUG: Authorization response keys: ${Object.keys(authResponse)}`
      );

      const uploadUrl = authResponse.uploadUrl;
      debugPrint(
        `DEBUG: Upload URL obtained: ${uploadUrl.substring(0, 50)}...`
      );

      debugPrint("DEBUG: Uploading file...");

      const fileStream = createReadStream(imagePath);
      const uploadResponse = await fetch(uploadUrl, {
        method: "PUT",
        body: fileStream,
        headers: {
          "x-amz-meta-nvcf-asset-description": description,
          "content-type": contentType,
        },
      });

      debugPrint(`DEBUG: Upload response status: ${uploadResponse.status}`);

      if (!uploadResponse.ok) {
        throw new Error(
          `Upload failed: ${uploadResponse.status} ${uploadResponse.statusText}`
        );
      }

      const assetId = authResponse.assetId;
      debugPrint(`DEBUG: Asset uploaded successfully, ID: ${assetId}`);
      return assetId;
    } catch (error) {
      debugPrint(`DEBUG: Request exception during upload: ${error}`);
      throw new Error(`Failed to upload asset: ${error}`);
    }
  }

  /**
   * Classify an image and return the result
   */
  async classifyImage(imagePath: string, task = "<CAPTION>"): Promise<string> {
    try {
      debugPrint(
        `DEBUG: Starting classification for ${imagePath} with task ${task}`
      );

      // Try direct base64 approach first
      try {
        return await this.classifyImageDirect(imagePath, task);
      } catch (directError) {
        debugPrint(
          `DEBUG: Direct method failed, trying asset upload: ${directError}`
        );
        // Fall back to asset upload method
        const assetId = await this.uploadAsset(imagePath);
        const content = `${task}<img src="data:image/jpeg;asset_id,${assetId}" />`;
        const inputs = {
          messages: [
            {
              role: "user",
              content,
            },
          ],
        };

        debugPrint(
          "DEBUG: Classification input prepared (asset upload method)"
        );
        return await this.sendClassificationRequest(inputs, assetId);
      }
    } catch (error) {
      debugPrint(`DEBUG: Classification exception: ${error}`);
      throw new Error(`Classification failed: ${error}`);
    }
  }

  /**
   * Classify image using direct base64 encoding
   */
  async classifyImageDirect(
    imagePath: string,
    task = "<CAPTION>"
  ): Promise<string> {
    debugPrint(`DEBUG: Trying direct base64 method for ${imagePath}`);

    // Read and encode image as base64
    const imageBuffer = await fs.readFile(imagePath);
    const base64Image = imageBuffer.toString("base64");

    // Determine content type
    let contentType = lookup(imagePath) || "image/jpeg";
    if (!contentType.startsWith("image/")) {
      contentType = "image/jpeg";
    }

    const content = `${task}<img src="data:${contentType};base64,${base64Image}" />`;
    const inputs = {
      messages: [
        {
          role: "user",
          content,
        },
      ],
    };

    debugPrint("DEBUG: Classification input prepared (direct base64 method)");
    return await this.sendClassificationRequest(inputs);
  }

  /**
   * Send classification request to API
   */
  async sendClassificationRequest(
    inputs: any,
    assetId?: string
  ): Promise<string> {
    const headers: any = {
      "Content-Type": "application/json",
      Authorization: this.headerAuth,
      Accept: "application/json",
    };

    // Only add asset headers if we have an assetId (for asset upload method)
    if (assetId) {
      headers["NVCF-INPUT-ASSET-REFERENCES"] = assetId;
      headers["NVCF-FUNCTION-ASSET-IDS"] = assetId;
    }

    debugPrint(`DEBUG: Sending classification request to ${this.apiUrl}`);

    const response = await fetch(this.apiUrl, {
      method: "POST",
      headers,
      body: JSON.stringify(inputs),
    });

    debugPrint(`DEBUG: Classification response status: ${response.status}`);

    if (!response.ok) {
      const errorText = await response.text();
      debugPrint(`DEBUG: Classification error response: ${errorText}`);
      throw new Error(
        `Classification request failed: ${response.status} ${response.statusText}. Response: ${errorText}`
      );
    }

    const contentType =
      response.headers.get("content-type")?.toLowerCase() || "";
    debugPrint(`DEBUG: Response content type: ${contentType}`);

    if (contentType.includes("application/json")) {
      const result = (await response.json()) as ClassificationResponse;
      debugPrint("DEBUG: JSON response received");
      return result.choices[0].message.content;
    } else if (contentType.includes("application/zip")) {
      debugPrint("DEBUG: ZIP response received, extracting...");
      const buffer = await response.buffer();
      return this.extractZipResponse(buffer);
    } else {
      debugPrint(`DEBUG: Unknown response format: ${contentType}`);
      return "Classification failed - unknown response format";
    }
  }

  /**
   * Extract classification result from ZIP response
   */
  private extractZipResponse(zipBuffer: Buffer): string {
    try {
      const zip = new AdmZip(zipBuffer);
      const entries = zip.getEntries();

      const responseEntry = entries.find((entry) =>
        entry.entryName.endsWith(".response")
      );

      if (responseEntry) {
        const jsonContent = JSON.parse(
          responseEntry.getData().toString("utf8")
        );
        let content = jsonContent.choices[0].message.content;

        if (content.startsWith("<CAPTION>")) {
          content = content.substring(9);
        }

        return content;
      }

      return "Classification failed - could not extract result from ZIP";
    } catch (error) {
      return `Classification failed - ZIP extraction error: ${error}`;
    }
  }
}

/**
 * Save classification result to JSON file
 */
export async function saveClassificationResult(
  imagePath: string,
  classificationResult: string,
  outputDir: string
): Promise<string> {
  try {
    await fs.mkdir(outputDir, { recursive: true });
  } catch (error) {
    // Directory might already exist
  }

  const imageFilename = path.basename(imagePath);
  const timestamp = new Date().toISOString();

  const resultData: ResultData = {
    timestamp,
    image_file: imageFilename,
    image_path: imagePath,
    classification: classificationResult,
    processed_at: timestamp,
  };

  const baseName = path.parse(imageFilename).name;
  const jsonFilename = `${baseName}_classification.json`;
  const jsonPath = path.join(outputDir, jsonFilename);

  await fs.writeFile(jsonPath, JSON.stringify(resultData, null, 2), "utf8");

  return jsonPath;
}

/**
 * Main function for command-line usage
 */
async function main(): Promise<void> {
  const args = process.argv.slice(2);

  if (args.length < 2) {
    console.error(
      "Usage: node imageClassifier.js <image_path> <output_dir> [--task <task>] [--debug]"
    );
    process.exit(1);
  }

  const imagePath = args[0];
  const outputDir = args[1];
  let task = "<CAPTION>";

  // Parse optional arguments
  for (let i = 2; i < args.length; i++) {
    if (args[i] === "--task" && i + 1 < args.length) {
      task = args[i + 1];
      i++;
    } else if (args[i] === "--debug") {
      DEBUG_MODE = true;
    }
  }

  try {
    // Debug logging
    debugPrint(
      `DEBUG: TypeScript script started with args: ${JSON.stringify(args)}`
    );
    debugPrint(`DEBUG: Current working directory: ${process.cwd()}`);
    debugPrint(`DEBUG: Node version: ${process.version}`);
    debugPrint(`DEBUG: Image path: ${imagePath}`);
    debugPrint(`DEBUG: Output directory: ${outputDir}`);

    // Check API key
    if (!API_KEY) {
      throw new Error(
        "API_KEY is not set. Check config.js or environment variables."
      );
    }
    debugPrint(`DEBUG: API key is present (length: ${API_KEY.length})`);

    // Validate inputs
    if (!existsSync(imagePath)) {
      throw new Error(`Image file not found: ${imagePath}`);
    }

    // Check file size
    const stats = statSync(imagePath);
    const fileSize = stats.size;
    debugPrint(`DEBUG: Image file size: ${fileSize} bytes`);

    if (fileSize === 0) {
      throw new Error("Image file is empty");
    }

    // Initialize classifier and process image
    debugPrint("DEBUG: Initializing classifier...");
    const classifier = new ImageClassifier();

    debugPrint("DEBUG: Starting classification...");
    const classification = await classifier.classifyImage(imagePath, task);
    debugPrint(
      `DEBUG: Classification completed: ${classification.substring(0, 100)}...`
    );

    const resultPath = await saveClassificationResult(
      imagePath,
      classification,
      outputDir
    );
    debugPrint(`DEBUG: Result saved to: ${resultPath}`);

    // Output result as JSON for Node.js to parse
    const result: ClassificationResult = {
      success: true,
      classification,
      result_file: resultPath,
      image_file: path.basename(imagePath),
    };

    console.log("CLASSIFICATION_RESULT:" + JSON.stringify(result));
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    const errorType =
      error instanceof Error ? error.constructor.name : "Unknown";

    console.error(`ERROR: ${errorMessage}`);
    console.error(`ERROR: Exception type: ${errorType}`);
    console.error(
      `ERROR: Stack trace:\n${
        error instanceof Error ? error.stack : "No stack trace available"
      }`
    );

    const result: ClassificationResult = {
      success: false,
      error: errorMessage,
      error_type: errorType,
      image_file: existsSync(imagePath) ? path.basename(imagePath) : "unknown",
    };

    console.log("CLASSIFICATION_RESULT:" + JSON.stringify(result));
    process.exit(1);
  }
}

// Set debug mode for module usage
export function setDebugMode(enabled: boolean): void {
  DEBUG_MODE = enabled;
}

// Run main function if this file is executed directly
if (require.main === module) {
  main().catch((error) => {
    console.error("Unhandled error:", error);
    process.exit(1);
  });
}
